File: go/patience/patience.go
// Patience Diff algorithm implementation.
//
// The algorithm finds lines that appear exactly once in each side,
// computes the Longest Increasing Subsequence (LIS) of their positions
// to establish anchors, then recursively diffs regions between anchors.
// Regions with no unique lines fall back to the provided fallback differ.

package patience

// Anchor represents a unique line matched between left and right.
type Anchor struct {
	LeftIdx  int
	RightIdx int
}

// DiffFunc computes diff ranges for sub-regions. Returns a list of
// (LeftIdx, RightIdx) pairs representing equal lines.
type DiffFunc func(left, right []string) []Anchor

// Diff runs the Patience Diff algorithm on left and right line slices.
// fallback is used for regions with no unique common lines.
// Returns a list of matched (equal) line pairs as anchors.
func Diff(left, right []string, fallback DiffFunc) []Anchor {
	var result []Anchor
	diffRecursive(left, right, 0, 0, fallback, &result)
	return result
}

func diffRecursive(left, right []string, offsetL, offsetR int, fallback DiffFunc, result *[]Anchor) {
	if len(left) == 0 || len(right) == 0 {
		return
	}

	anchors := findAnchors(left, right)

	if len(anchors) == 0 {
		// No unique common lines — use fallback (Myers)
		matches := fallback(left, right)
		for _, m := range matches {
			*result = append(*result, Anchor{LeftIdx: m.LeftIdx + offsetL, RightIdx: m.RightIdx + offsetR})
		}
		return
	}

	// Process regions between anchors
	prevL, prevR := 0, 0
	for _, a := range anchors {
		if a.LeftIdx > prevL || a.RightIdx > prevR {
			diffRecursive(
				left[prevL:a.LeftIdx], right[prevR:a.RightIdx],
				offsetL+prevL, offsetR+prevR,
				fallback, result,
			)
		}
		*result = append(*result, Anchor{LeftIdx: a.LeftIdx + offsetL, RightIdx: a.RightIdx + offsetR})
		prevL = a.LeftIdx + 1
		prevR = a.RightIdx + 1
	}

	// Diff the region after the last anchor
	if prevL < len(left) || prevR < len(right) {
		diffRecursive(
			left[prevL:], right[prevR:],
			offsetL+prevL, offsetR+prevR,
			fallback, result,
		)
	}
}

// findAnchors finds lines unique in both sides, matches them, and returns
// the LIS of their right-side indices as anchors.
func findAnchors(left, right []string) []Anchor {
	leftCount := make(map[string]int)
	for _, l := range left {
		leftCount[l]++
	}

	rightCount := make(map[string]int)
	for _, r := range right {
		rightCount[r]++
	}

	// Build index of unique lines in right
	rightIndex := make(map[string]int)
	for i, r := range right {
		if rightCount[r] == 1 {
			rightIndex[r] = i
		}
	}

	// Collect matching pairs: lines unique in both sides
	var pairs []Anchor
	for i, l := range left {
		if leftCount[l] == 1 {
			if rIdx, ok := rightIndex[l]; ok && rightCount[l] == 1 {
				pairs = append(pairs, Anchor{LeftIdx: i, RightIdx: rIdx})
			}
		}
	}

	if len(pairs) == 0 {
		return nil
	}

	return lis(pairs)
}

// lis computes the Longest Increasing Subsequence of anchors by RightIdx.
// Input must be sorted by LeftIdx. Uses patience sorting (O(n log n)).
func lis(pairs []Anchor) []Anchor {
	n := len(pairs)
	if n == 0 {
		return nil
	}

	// tails[i] = index into pairs of the smallest RightIdx ending an IS of length i+1
	tails := make([]int, 0, n)
	// prev[i] = index of previous element in the LIS ending at pairs[i]
	prev := make([]int, n)
	for i := range prev {
		prev[i] = -1
	}

	for i, p := range pairs {
		lo, hi := 0, len(tails)
		for lo < hi {
			mid := (lo + hi) / 2
			if pairs[tails[mid]].RightIdx < p.RightIdx {
				lo = mid + 1
			} else {
				hi = mid
			}
		}

		if lo == len(tails) {
			tails = append(tails, i)
		} else {
			tails[lo] = i
		}

		if lo > 0 {
			prev[i] = tails[lo-1]
		}
	}

	length := len(tails)
	result := make([]Anchor, length)
	idx := tails[length-1]
	for i := length - 1; i >= 0; i-- {
		result[i] = pairs[idx]
		idx = prev[idx]
	}
	return result
}

File: swift/Sources/FileDiff/Internal/Patience.swift
// Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
// Ported from intellij-community to Swift (via the Go port in this repository).

/// A matched pair of line indices produced by the Patience diff algorithm.
struct PatienceAnchor: Equatable {
    let leftIndex: Int
    let rightIndex: Int
}

/// Fallback matcher used for sub-regions that contain no unique common lines.
typealias PatienceFallback = @Sendable (_ left: [String], _ right: [String]) -> [PatienceAnchor]

/// Runs Patience Diff on two line sequences.
///
/// Anchors are unique common lines whose right-side positions form a longest
/// increasing subsequence. Sub-regions between anchors are recursively diffed,
/// and regions without unique anchors delegate to `fallback`.
func patienceDiff(
    _ left: [String], _ right: [String], fallback: PatienceFallback
) -> [PatienceAnchor] {
    var result: [PatienceAnchor] = []
    patienceDiffRecursive(left, right, offsetL: 0, offsetR: 0, fallback: fallback, result: &result)
    return result
}

// swiftlint:disable:next function_parameter_count
private func patienceDiffRecursive(
    _ left: [String], _ right: [String],
    offsetL: Int, offsetR: Int,
    fallback: PatienceFallback,
    result: inout [PatienceAnchor]
) {
    if left.isEmpty || right.isEmpty {
        return
    }
    let anchors = findAnchors(left, right)

    if anchors.isEmpty {
        let matches = fallback(left, right)
        for match in matches {
            result.append(PatienceAnchor(
                leftIndex: match.leftIndex + offsetL,
                rightIndex: match.rightIndex + offsetR
            ))
        }
        return
    }

    var prevL = 0
    var prevR = 0
    for anchor in anchors {
        if anchor.leftIndex > prevL || anchor.rightIndex > prevR {
            patienceDiffRecursive(
                Array(left[prevL..<anchor.leftIndex]),
                Array(right[prevR..<anchor.rightIndex]),
                offsetL: offsetL + prevL, offsetR: offsetR + prevR,
                fallback: fallback, result: &result
            )
        }
        result.append(PatienceAnchor(
            leftIndex: anchor.leftIndex + offsetL,
            rightIndex: anchor.rightIndex + offsetR
        ))
        prevL = anchor.leftIndex + 1
        prevR = anchor.rightIndex + 1
    }

    if prevL < left.count || prevR < right.count {
        patienceDiffRecursive(
            Array(left[prevL...]),
            Array(right[prevR...]),
            offsetL: offsetL + prevL, offsetR: offsetR + prevR,
            fallback: fallback, result: &result
        )
    }
}

private func findAnchors(_ left: [String], _ right: [String]) -> [PatienceAnchor] {
    var leftCount: [String: Int] = [:]
    for line in left {
        leftCount[line, default: 0] += 1
    }
    var rightCount: [String: Int] = [:]
    for line in right {
        rightCount[line, default: 0] += 1
    }
    var rightIndex: [String: Int] = [:]
    for (index, line) in right.enumerated() where rightCount[line] == 1 {
        rightIndex[line] = index
    }

    var pairs: [PatienceAnchor] = []
    for (index, line) in left.enumerated() where leftCount[line] == 1 {
        if rightCount[line] == 1, let rIdx = rightIndex[line] {
            pairs.append(PatienceAnchor(leftIndex: index, rightIndex: rIdx))
        }
    }
    if pairs.isEmpty {
        return []
    }
    return longestIncreasingSubsequence(pairs)
}

/// Longest increasing subsequence of `pairs` by `rightIndex`.
///
/// Input must be sorted by `leftIndex`. Uses patience sorting (`O(n log n)`).
func longestIncreasingSubsequence(_ pairs: [PatienceAnchor]) -> [PatienceAnchor] {
    let count = pairs.count
    if count == 0 {
        return []
    }

    var tails: [Int] = []
    tails.reserveCapacity(count)
    var prev = Array(repeating: -1, count: count)

    for (index, pair) in pairs.enumerated() {
        var lo = 0
        var hi = tails.count
        while lo < hi {
            let mid = (lo + hi) / 2
            if pairs[tails[mid]].rightIndex < pair.rightIndex {
                lo = mid + 1
            } else {
                hi = mid
            }
        }
        if lo == tails.count {
            tails.append(index)
        } else {
            tails[lo] = index
        }
        if lo > 0 {
            prev[index] = tails[lo - 1]
        }
    }

    var result = Array(repeating: PatienceAnchor(leftIndex: 0, rightIndex: 0), count: tails.count)
    var idx = tails[tails.count - 1]
    for position in (0..<tails.count).reversed() {
        result[position] = pairs[idx]
        idx = prev[idx]
    }
    return result
}
